File: src/main/java/no/novari/flyt/archive/gateway/resource/configuration/FintLinksResourcePipelineFactory.java
package no.novari.flyt.archive.gateway.resource.configuration;

import no.fint.model.resource.FintLinks;
import no.novari.cache.FintCache;
import no.novari.flyt.archive.gateway.links.ResourceLinkUtil;
import no.novari.kafka.topic.name.EntityTopicNameParameters;
import no.novari.kafka.topic.name.TopicNamePrefixParameters;
import org.springframework.stereotype.Service;

import java.util.List;

@Service
public class FintLinksResourcePipelineFactory {
    public <T extends FintLinks> ResourcePipeline<T> createResourcePipeLine(
            Class<T> resourceClass,
            FintCache<String, T> cache,
            List<String> resourceReferencePath,
            boolean publishOnKafka
    ) {
        ResourcePipeline.ResourcePipelineBuilder<T> resourcePipelineBuilder = ResourcePipeline
                .<T>builder()
                .resourceClass(resourceClass)
                .urlResourcePath(String.join("/", resourceReferencePath))
                .cacheProperties(
                        ResourcePipelineCacheProperties
                                .<T>builder()
                                .createKeys(ResourceLinkUtil::getSelfLinks)
                                .cache(cache)
                                .build()
                );

        if (publishOnKafka) {
            resourcePipelineBuilder.kafkaProperties(
                    ResourcePipelineKafkaProperties
                            .<T>builder()
                            .createKafkaKey(ResourceLinkUtil::getFirstSelfLink)
                            .topicNameParameters(
                                    EntityTopicNameParameters
                                            .builder()
                                            .topicNamePrefixParameters(TopicNamePrefixParameters
                                                    .stepBuilder()
                                                    .orgIdApplicationDefault()
                                                    .domainContextApplicationDefault()
                                                    .build()
                                            )
                                            .resourceName(String.join("-", resourceReferencePath))
                                            .build()
                            )
                            .build()
            );
        }

        return resourcePipelineBuilder.build();
    }
}

File: src/main/kotlin/no/novari/flyt/archive/gateway/resource/configuration/FintLinksResourcePipelineFactory.kt
package no.novari.flyt.archive.gateway.resource.configuration

import no.novari.cache.FintCache
import no.novari.fint.model.resource.FintLinks
import no.novari.flyt.archive.gateway.links.ResourceLinkUtil
import no.novari.kafka.topic.name.EntityTopicNameParameters
import no.novari.kafka.topic.name.TopicNamePrefixParameters
import org.springframework.stereotype.Service

@Service
class FintLinksResourcePipelineFactory {
    fun <T : FintLinks> createResourcePipeLine(
        resourceClass: Class<T>,
        cache: FintCache<String, T>,
        resourceReferencePath: List<String>,
        publishOnKafka: Boolean,
    ): ResourcePipeline<T> {
        val resourcePipelineBuilder =
            ResourcePipeline
                .builder<T>()
                .resourceClass(resourceClass)
                .urlResourcePath(resourceReferencePath.joinToString("/"))
                .cacheProperties(
                    ResourcePipelineCacheProperties
                        .builder<T>()
                        .createKeys(ResourceLinkUtil::getSelfLinks)
                        .cache(cache)
                        .build(),
                )

        if (publishOnKafka) {
            resourcePipelineBuilder.kafkaProperties(
                ResourcePipelineKafkaProperties
                    .builder<T>()
                    .createKafkaKey(ResourceLinkUtil::getFirstSelfLink)
                    .topicNameParameters(
                        EntityTopicNameParameters
                            .builder()
                            .topicNamePrefixParameters(
                                TopicNamePrefixParameters
                                    .stepBuilder()
                                    .orgIdApplicationDefault()
                                    .domainContextApplicationDefault()
                                    .build(),
                            ).resourceName(resourceReferencePath.joinToString("-"))
                            .build(),
                    ).build(),
            )
        }

        return resourcePipelineBuilder.build()
    }
}
